evm_arithmetization/src/verifier.rs

```diff
@@ -15,9 +20,75 @@ use crate::cpu::kernel::aggregator::KERNEL;
 use crate::cpu::kernel::constants::global_metadata::GlobalMetadata;
 use crate::memory::segments::Segment;
 use crate::memory::VALUE_LIMBS;
-use crate::proof::{AllProof, AllProofChallenges, PublicValues};
+use crate::proof::{AllProof, AllProofChallenges, MemCap, PublicValues};
 use crate::util::h2u;
 
+pub(crate) fn initial_memory_merkle_cap<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    rate_bits: usize,
+    cap_height: usize,
+) -> MerkleCap<F, C::Hasher> {
+    // At the start of a transaction proof, `MemBefore` only contains the RLP
+    // constant and the `ShiftTable`.
+    let mut trace = vec![];
+
+    // Push kernel code.
+    for (i, &byte) in KERNEL.code.iter().enumerate() {
+        let mut row = vec![F::ZERO; crate::memory_continuation::columns::NUM_COLUMNS];
+        row[crate::memory_continuation::columns::FILTER] = F::ONE;
+        row[crate::memory_continuation::columns::ADDR_CONTEXT] = F::ZERO;
+        row[crate::memory_continuation::columns::ADDR_SEGMENT] =
+            F::from_canonical_usize(Segment::Code.unscale());
+        row[crate::memory_continuation::columns::ADDR_VIRTUAL] = F::from_canonical_usize(i);
+        row[crate::memory_continuation::columns::value_limb(0)] = F::from_canonical_u8(byte);
+        trace.push(row);
+    }
```

Comment:
Would be nice to not have this code duplication (same for shift table below) with `fixed_recursive_verifier.rs`

Reply:
Yup, forgot to remove it, thanks!